### PR TITLE
Add peak memory distribution among tasks of each stage

### PR DESCRIPTION
**Pull request by viczhang861**
Resolves https://github.com/prestodb/presto/issues/13327

```
== RELEASE NOTES ==

General Changes
* Add peak total memory distribution among tasks of each stage

```

### CODE REVIEW

File: presto-spi/src/main/java/com/facebook/presto/spi/eventlistener/TaskDistribution.java [hunk 1]
```diff
@@ -44,7 +44,7 @@ public StageCpuDistribution(
             @JsonProperty("min") long min,
             @JsonProperty("max") long max,
             @JsonProperty("total") long total,
-            @JsonProperty("average") double average)
+            @JsonProperty("average") float average)
```
Comment (aweisberg) [inline presto-spi/src/main/java/com/facebook/presto/spi/eventlistener/TaskDistribution.java]: Just curious, why reduce precision?

File: presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java [hunk 2]
```diff
@@ -507,49 +516,53 @@ private static void logQueryTimeline(
...
-                snapshot.getP50(),
-                snapshot.getP75(),
-                snapshot.getP90(),
-                snapshot.getP95(),
-                snapshot.getP99(),
-                snapshot.getMin(),
-                snapshot.getMax(),
-                (long) snapshot.getTotal(),
-                snapshot.getTotal() / snapshot.getCount());
+                cpuSnapshot.getP25(),
```
Comment (aweisberg) [inline presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java]: Constructing a task distribution is very verbose and a lot of this stuff comes from a DistributionSnapshot, can you update the TaskDistributionConstructor to reduce some of this duplication by working on the DistributionSnapshot directly?

Comment (viczhang861) [reply presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java]: This is intended to support serialization into json string. There is downstream dependency on structure of class.

File: presto-spi/src/main/java/com/facebook/presto/spi/eventlistener/TaskDistribution.java (hunk 1 above)
Comment (viczhang861) [reply presto-spi/src/main/java/com/facebook/presto/spi/eventlistener/TaskDistribution.java]: For current purpose of using this class for either memory bytes or cpu time,  float precision is enough.

File: presto-main/src/main/java/com/facebook/presto/operator/TaskStats.java [hunk 3]
```diff
@@ -137,6 +139,7 @@ public TaskStats(
             @JsonProperty("userMemoryReservation") DataSize userMemoryReservation,
             @JsonProperty("revocableMemoryReservation") DataSize revocableMemoryReservation,
             @JsonProperty("systemMemoryReservation") DataSize systemMemoryReservation,
+            @JsonProperty("peakTotalMemoryInBytes") long peakTotalMemoryInBytes,
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/TaskStats.java]: Looking at cumulativeUserMemory, I'd expect

`@JsonProperty("peakTotalMemory") double peakTotalMemory,`

File: presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java [hunk 4]
```diff
@@ -43,7 +43,7 @@
 import com.facebook.presto.spi.eventlistener.QueryMetadata;
 import com.facebook.presto.spi.eventlistener.QueryOutputMetadata;
 import com.facebook.presto.spi.eventlistener.QueryStatistics;
-import com.facebook.presto.spi.eventlistener.StageCpuDistribution;
+import com.facebook.presto.spi.eventlistener.TaskDistribution;
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java]: I'm wondering if it would be useful to keep `Stage` in the name. The data is the distribution of task resource usage within a stage, right? `StageResourceDistribution` maybe?

File: presto-spi/src/main/java/com/facebook/presto/spi/eventlistener/TaskDistribution.java (hunk 1 above)
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/eventlistener/TaskDistribution.java]: This is unrelated change and therefore should go into separate commit. My preference would be to keep `double` though. Primarily to avoid having other readers wonder why float.

File: presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java [hunk 5]
```diff
@@ -507,49 +516,53 @@ private static void logQueryTimeline(
...
-    private static void populateDistribution(StageInfo stageInfo, ImmutableList.Builder<TaskDistribution> distributions)
-    {
-        distributions.add(computeCpuDistribution(stageInfo));
-        for (StageInfo subStage : stageInfo.getSubStages()) {
-            populateDistribution(subStage, distributions);
-        }
-    }
-
-    private static TaskDistribution computeCpuDistribution(StageInfo stageInfo)
+    private static void computeTaskDistributions(
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java]: nit: perhaps, computeResourceDistributions or computeCpuAndMemoryDistributions

**Comment by oerling** [comment]
This is all fine by me. I tested it and the data looks right.

File: presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java [hunk 6]
```diff
@@ -96,6 +96,7 @@
 
     private final Object cumulativeMemoryLock = new Object();
     private final AtomicDouble cumulativeUserMemory = new AtomicDouble(0.0);
+    private final AtomicLong peakTotalMemoryInBytes = new AtomicLong(0);
```
[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java]: - cumulativeUserMemory uses double as its unit is Bytes * Time,   this explains the difference of names

[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java]: @viczhang861 Vic, thanks for explaining. It makes sense now.

File: presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java (hunk 5 above)
Comment (viczhang861) [reply presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java]: Done

File: presto-spi/src/main/java/com/facebook/presto/spi/eventlistener/TaskDistribution.java (hunk 1 above)
Comment (viczhang861) [reply presto-spi/src/main/java/com/facebook/presto/spi/eventlistener/TaskDistribution.java]: Done

File: presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java (hunk 4 above)
Comment (viczhang861) [reply presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java]: Updated to ResourceDistribution to avoid confusion

File: presto-spi/src/main/java/com/facebook/presto/spi/eventlistener/TaskDistribution.java (hunk 1 above)
Comment (viczhang861) [reply presto-spi/src/main/java/com/facebook/presto/spi/eventlistener/TaskDistribution.java]: Changed it back `double` to avoid confusion

File: presto-main/src/main/java/com/facebook/presto/operator/TaskStats.java (hunk 3 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/operator/TaskStats.java]: Or `peakTotalMemoryReservation` and use `DataSize` as type? No strong opinion here.

File: presto-spi/src/main/java/com/facebook/presto/spi/eventlistener/QueryStatistics.java [hunk 7]
```diff
@@ -73,6 +74,7 @@ public QueryStatistics(
             int completedSplits,
             boolean complete,
             List<ResourceDistribution> cpuTimeDistribution,
+            List<ResourceDistribution> peakMemoryDistribution,
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/eventlistener/QueryStatistics.java]: What about `taskPeakMemoryDistribution`? No strong opinion since `peakMemoryDistribution` is parallel to `cpuTimeDistribution`
